File: app/vitals.py
"""Host-wide CPU / memory / disk / network vitals via psutil.

Kanshi runs with `network_mode: host` and without lxcfs, so /proc/stat,
/proc/meminfo, /proc/diskstats and /proc/net/dev all report real host values
and psutil needs no special configuration.
"""
from __future__ import annotations

import os
import time

import psutil

from .config import config

_last_net: tuple[float, int, int] | None = None
_last_disk: tuple[float, int, int] | None = None
_last_rates: dict[str, tuple[float, float]] = {"net": (0.0, 0.0), "disk": (0.0, 0.0)}
_last_cpu_call = 0.0

# Below this gap the counter deltas are too small to divide by: a 20ms window
# turns a routine 2MB read into a fake 100MB/s spike. Hold the previous rate.
MIN_DT = 0.5


def prime() -> None:
    """Seed psutil's internal counters so the first sample isn't all zeros."""
    global _last_cpu_call
    psutil.cpu_percent(percpu=True, interval=None)
    _last_cpu_call = time.monotonic()
    _rate_net()
    _rate_disk()


def _cpu_cores() -> list[float]:
    """Per-core utilisation since the previous call.

    psutil's interval=None form is a delta against the last call, so if that
    call was moments ago every core reads 0%. When the gap is too short to be
    meaningful, measure a real (short, blocking) window instead — sample() runs
    in a worker thread, so this never stalls the event loop.
    """
    global _last_cpu_call
    gap = time.monotonic() - _last_cpu_call
    cores = psutil.cpu_percent(percpu=True, interval=None if gap >= MIN_DT else 0.25)
    _last_cpu_call = time.monotonic()
    return cores


def _rate_net() -> tuple[float, float]:
    global _last_net
    now = time.monotonic()
    try:
        c = psutil.net_io_counters()
    except Exception:
        return (0.0, 0.0)
    prev = _last_net
    if prev is not None and now - prev[0] < MIN_DT:
        return _last_rates["net"]
    _last_net = (now, c.bytes_recv, c.bytes_sent)
    if prev is None:
        return (0.0, 0.0)
    dt = now - prev[0]
    # Counters are 64-bit but can reset if an interface disappears; clamp.
    rates = (max(0, c.bytes_recv - prev[1]) / dt, max(0, c.bytes_sent - prev[2]) / dt)
    _last_rates["net"] = rates
    return rates


def _rate_disk() -> tuple[float, float]:
    global _last_disk
    now = time.monotonic()
    try:
        c = psutil.disk_io_counters()
    except Exception:
        return (0.0, 0.0)
    if c is None:
        return (0.0, 0.0)
    prev = _last_disk
    if prev is not None and now - prev[0] < MIN_DT:
        return _last_rates["disk"]
    _last_disk = (now, c.read_bytes, c.write_bytes)
    if prev is None:
        return (0.0, 0.0)
    dt = now - prev[0]
    rates = (max(0, c.read_bytes - prev[1]) / dt, max(0, c.write_bytes - prev[2]) / dt)
    _last_rates["disk"] = rates
    return rates


def _temperature() -> float | None:
    """Best-effort package temperature; needs /sys mounted (Docker does by default)."""
    try:
        temps = psutil.sensors_temperatures()
    except Exception:
        return None
    for key in ("coretemp", "k10temp", "cpu_thermal", "acpitz", "zenpower"):
        for entry in temps.get(key, []):
            if entry.current:
                return round(entry.current, 1)
    for entries in temps.values():
        for entry in entries:
            if entry.current:
                return round(entry.current, 1)
    return None


def filesystems() -> list[dict]:
    """Usage for each configured storage root, read straight off statvfs."""
    out = []
    for label, path in config.roots():
        try:
            st = os.statvfs(path)
        except OSError:
            continue
        total = st.f_blocks * st.f_frsize
        # f_bavail excludes root-reserved blocks, so used+free won't equal total.
        # Report "used" the way df does, against the non-reserved capacity.
        free = st.f_bavail * st.f_frsize
        used = total - st.f_bfree * st.f_frsize
        if total <= 0:
            continue
        out.append({
            "label": label,
            "path": path,
            "total": total,
            "used": used,
            "free": free,
            "percent": round(used / (used + free) * 100, 1) if used + free else 0.0,
        })
    return out


def sample() -> dict:
    per_core = _cpu_cores()
    mem = psutil.virtual_memory()
    swap = psutil.swap_memory()
    rx, tx = _rate_net()
    read, write = _rate_disk()
    try:
        load1, load5, load15 = os.getloadavg()
    except OSError:
        load1 = load5 = load15 = 0.0

    return {
        "ts": time.time(),
        "cpu": {
            "percent": round(sum(per_core) / len(per_core), 1) if per_core else 0.0,
            "cores": [round(c, 1) for c in per_core],
            "load": [round(load1, 2), round(load5, 2), round(load15, 2)],
            "count": len(per_core),
            "temp": _temperature(),
        },
        "memory": {
            "total": mem.total,
            "used": mem.total - mem.available,
            "available": mem.available,
            "cached": getattr(mem, "cached", 0) + getattr(mem, "buffers", 0),
            "percent": round((mem.total - mem.available) / mem.total * 100, 1) if mem.total else 0.0,
        },
        "swap": {
            "total": swap.total,
            "used": swap.used,
            "percent": round(swap.percent, 1),
        },
        "network": {"rx": rx, "tx": tx},
        "diskio": {"read": read, "write": write},
        "filesystems": filesystems(),
        "uptime": time.time() - psutil.boot_time(),
    }

File: internal/vitals/vitals.go
// Package vitals reads host-wide CPU / memory / disk / network figures.
//
// On Linux they come straight out of /proc and /sys. Kanshi runs with
// `network_mode: host` and without lxcfs, so those report real host values
// with no special configuration. On Windows they come from the same kernel32,
// ntdll and iphlpapi calls Task Manager uses.
package vitals

import (
	"math"
	"sync"
	"time"

	"github.com/rene-roid/kanshi/internal/roots"
)

// Below this gap the counter deltas are too small to divide by: a 20ms window
// turns a routine 2MB read into a fake 100MB/s spike. Hold the previous rate.
const minDT = 500 * time.Millisecond

// Reader owns the counter baselines every rate is measured against. One is
// created per process; the mutex only ever guards against a REST handler
// sampling at the same moment as the poller.
type Reader struct {
	roots *roots.Resolver

	mu       sync.Mutex
	sys      sysState // the platform's own baselines and cached handles
	prevNet  counterPair
	prevDisk counterPair
	netRate  [2]float64
	diskRate [2]float64
}

type counterPair struct {
	at   time.Time
	a, b uint64
	ok   bool
}

func New(r *roots.Resolver) *Reader { return &Reader{roots: r} }

/* ── payload ────────────────────────────────────────────────────────────── */

// Sample is the JSON the browser consumes; field names are part of the wire
// contract with web/app.js.
type Sample struct {
	TS          float64      `json:"ts"`
	CPU         CPU          `json:"cpu"`
	Memory      Memory       `json:"memory"`
	Swap        Swap         `json:"swap"`
	Network     RxTx         `json:"network"`
	DiskIO      ReadWrite    `json:"diskio"`
	Filesystems []Filesystem `json:"filesystems"`
	Uptime      float64      `json:"uptime"`
}

type CPU struct {
	Percent float64   `json:"percent"`
	Cores   []float64 `json:"cores"`
	// Load average and temperature are null where the platform has no such
	// thing (Windows has no load average) or no readable sensor.
	Load  *[3]float64 `json:"load"`
	Count int         `json:"count"`
	Temp  *float64    `json:"temp"`
}

type Memory struct {
	Total     uint64  `json:"total"`
	Used      uint64  `json:"used"`
	Available uint64  `json:"available"`
	Cached    uint64  `json:"cached"`
	Percent   float64 `json:"percent"`
}

type Swap struct {
	Total   uint64  `json:"total"`
	Used    uint64  `json:"used"`
	Percent float64 `json:"percent"`
}

type RxTx struct {
	RX float64 `json:"rx"`
	TX float64 `json:"tx"`
}

type ReadWrite struct {
	Read  float64 `json:"read"`
	Write float64 `json:"write"`
}

type Filesystem struct {
	Label   string  `json:"label"`
	Path    string  `json:"path"`
	Total   uint64  `json:"total"`
	Used    uint64  `json:"used"`
	Free    uint64  `json:"free"`
	Percent float64 `json:"percent"`
}

/* ── entry points ───────────────────────────────────────────────────────── */

// Prime seeds the counter baselines so the first published frame shows real
// numbers instead of a screen of zeros.
func (r *Reader) Prime() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.primeCPU()
	rx, tx, ok := r.netCounters()
	rate(&r.prevNet, &r.netRate, rx, tx, ok)
	read, write, ok := r.diskCounters()
	rate(&r.prevDisk, &r.diskRate, read, write, ok)
}

// Sample takes one full reading of the host.
func (r *Reader) Sample() Sample {
	r.mu.Lock()
	defer r.mu.Unlock()

	cores := r.cpuCores()
	mem, sw := r.memory()

	rx, tx, netOK := r.netCounters()
	net := rate(&r.prevNet, &r.netRate, rx, tx, netOK)
	read, write, diskOK := r.diskCounters()
	disk := rate(&r.prevDisk, &r.diskRate, read, write, diskOK)

	var mean float64
	if len(cores) > 0 {
		var sum float64
		for _, c := range cores {
			sum += c
		}
		mean = round1(sum / float64(len(cores)))
	}
	if cores == nil {
		cores = []float64{} // the browser iterates this; never ship null
	}

	return Sample{
		TS: float64(time.Now().UnixNano()) / 1e9,
		CPU: CPU{
			Percent: mean,
			Cores:   cores,
			Load:    r.loadAvg(),
			Count:   len(cores),
			Temp:    r.temperature(),
		},
		Memory:      mem,
		Swap:        sw,
		Network:     RxTx{RX: net[0], TX: net[1]},
		DiskIO:      ReadWrite{Read: disk[0], Write: disk[1]},
		Filesystems: r.Filesystems(),
		Uptime:      r.uptime(),
	}
}

// Filesystems reports usage for each storage root's volume. Roots that share a
// volume — C:\ and the profile folder on it — get one meter, labelled with the
// first of them.
func (r *Reader) Filesystems() []Filesystem {
	list := r.roots.Roots()
	out := make([]Filesystem, 0, len(list))
	seen := make(map[string]bool, len(list))
	for _, root := range list {
		u, err := roots.DiskUsage(root.Path)
		if err != nil || u.Total == 0 || seen[u.Volume] {
			continue
		}
		seen[u.Volume] = true
		fs := Filesystem{Label: root.Label, Path: root.Path, Total: u.Total, Used: u.Used, Free: u.Free}
		if u.Used+u.Free > 0 {
			fs.Percent = round1(float64(u.Used) / float64(u.Used+u.Free) * 100)
		}
		out = append(out, fs)
	}
	return out
}

/* ── shared arithmetic ──────────────────────────────────────────────────── */

// busyPercent turns two counter deltas into a utilisation, clamped to 0–100.
func busyPercent(busyDelta, allDelta float64) float64 {
	if allDelta <= 0 || busyDelta <= 0 {
		return 0
	}
	return round1(math.Min(busyDelta/allDelta*100, 100))
}

func memoryFigures(total, avail, cached uint64) Memory {
	if avail > total {
		avail = total
	}
	used := total - avail
	m := Memory{Total: total, Used: used, Available: avail, Cached: cached}
	if total > 0 {
		m.Percent = round1(float64(used) / float64(total) * 100)
	}
	return m
}

func swapFigures(total, free uint64) Swap {
	if free > total {
		free = total
	}
	s := Swap{Total: total, Used: total - free}
	if total > 0 {
		s.Percent = round1(float64(s.Used) / float64(total) * 100)
	}
	return s
}

// rate turns two counter readings into bytes/second, holding the previous
// answer when the window is too short to divide by.
func rate(prev *counterPair, held *[2]float64, a, b uint64, ok bool) [2]float64 {
	if !ok {
		return [2]float64{0, 0}
	}
	now := time.Now()
	if prev.ok && now.Sub(prev.at) < minDT {
		return *held
	}
	was := *prev
	*prev = counterPair{at: now, a: a, b: b, ok: true}
	if !was.ok {
		return [2]float64{0, 0}
	}
	dt := now.Sub(was.at).Seconds()
	if dt <= 0 {
		return *held
	}
	// Counters are 64-bit but reset when an interface or disk disappears;
	// clamp rather than reporting a negative or an absurd spike.
	*held = [2]float64{diff(a, was.a) / dt, diff(b, was.b) / dt}
	return *held
}

func diff(now, prev uint64) float64 {
	if now < prev {
		return 0
	}
	return float64(now - prev)
}

func round1(v float64) float64 { return math.Round(v*10) / 10 }
func round2(v float64) float64 { return math.Round(v*100) / 100 }
